File: app/src/main/java/com/wlp/myanjunote/customview/sample/easy/MainActivity.java
package com.wlp.myanjunote.customview.sample.easy;

import android.annotation.SuppressLint;
import android.content.Intent;
import android.os.Bundle;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.TextView;

import androidx.annotation.NonNull;
import androidx.appcompat.app.AppCompatActivity;
import androidx.recyclerview.widget.LinearLayoutManager;
import androidx.recyclerview.widget.RecyclerView;

import com.wlp.myanjunote.R;
import com.wlp.myanjunote.customview.sample.easy.activity.BarAndLineActivity;
import com.wlp.myanjunote.customview.sample.easy.activity.BarChartActivity;
import com.wlp.myanjunote.customview.sample.easy.activity.CombineChartActivity;
import com.wlp.myanjunote.customview.sample.easy.activity.DoubleBarCharActivity;
import com.wlp.myanjunote.customview.sample.easy.activity.HollowPieChartActivity;
import com.wlp.myanjunote.customview.sample.easy.activity.HollowPieChartNewActivity;
import com.wlp.myanjunote.customview.sample.easy.activity.LineChartActivity;
import com.wlp.myanjunote.customview.sample.easy.activity.PercentActivity;
import com.wlp.myanjunote.customview.sample.easy.activity.PieChartActivity;
import com.wlp.myanjunote.customview.sample.easy.activity.ScaleActivity;
import com.wlp.myanjunote.customview.sample.easy.entity.MainBean;

import java.util.ArrayList;
import java.util.List;

public class MainActivity extends AppCompatActivity {
    private RecyclerView mRecyclerView;
    private List<MainBean> mList = new ArrayList<>();

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_main4);
        initData();
        setRecyclerView();
    }

    private void initData() {
        mList.add(new MainBean("线性图", new Intent(this, LineChartActivity.class)));
        mList.add(new MainBean("柱状图", new Intent(this, BarChartActivity.class)));
        mList.add(new MainBean("双柱状图", new Intent(this, DoubleBarCharActivity.class)));
        mList.add(new MainBean("饼状图", new Intent(this, PieChartActivity.class)));
        mList.add(new MainBean("空心饼", new Intent(this, HollowPieChartActivity.class)));
        mList.add(new MainBean("空心饼1", new Intent(this, HollowPieChartNewActivity.class)));
        mList.add(new MainBean("直线比例", new Intent(this, ScaleActivity.class)));
        mList.add(new MainBean("组合图", new Intent(this, CombineChartActivity.class)));
        mList.add(new MainBean("线和柱", new Intent(this, BarAndLineActivity.class)));
        mList.add(new MainBean("比例图", new Intent(this, PercentActivity.class)));
    }

    private void setRecyclerView() {
        mRecyclerView = findViewById(R.id.recyclerview);
        mRecyclerView.setLayoutManager(new LinearLayoutManager(this));
        MyAdapter adapter = new MyAdapter(mList);
        adapter.setOnItemClickListener(position -> {
            startActivity(mList.get(position).getIntent());
        });
        mRecyclerView.setAdapter(adapter);
    }

    class MyAdapter extends RecyclerView.Adapter<MyAdapter.ViewHolder> {
        private List<MainBean> list = new ArrayList<>();
        private OnItemClickListener mListener;

        public MyAdapter(List<MainBean> data) {
            list = data;
        }

        public void setOnItemClickListener(OnItemClickListener listener) {
            mListener = listener;
        }

        @Override
        public ViewHolder onCreateViewHolder(ViewGroup parent, int viewType) {
            View view = LayoutInflater.from(parent.getContext()).inflate(R.layout.item_main, parent, false);
            return new ViewHolder(view);
        }

        @Override
        public void onBindViewHolder(@NonNull ViewHolder holder, @SuppressLint("RecyclerView") int position) {
            holder.mTextView.setText(list.get(position).getName());
            holder.itemView.setOnClickListener(new View.OnClickListener() {
                @Override
                public void onClick(View v) {
                    if (mListener != null) {
                        mListener.onItemClick(position);
                    }
                }
            });
        }

        @Override
        public int getItemCount() {
            return list.size();
        }

        class ViewHolder extends RecyclerView.ViewHolder {
            TextView mTextView;

            public ViewHolder(View itemView) {
                super(itemView);
                mTextView = itemView.findViewById(R.id.tv_name);
            }
        }
    }

    interface OnItemClickListener {
        void onItemClick(int position);
    }
}

File: app/src/main/java/com/wlp/myanjunote/customview/bitmapanddrawable/MainActivity.kt
package com.wlp.myanjunote.customview.bitmapanddrawable

import android.graphics.Bitmap
import android.graphics.drawable.BitmapDrawable
import android.graphics.drawable.ColorDrawable
import androidx.appcompat.app.AppCompatActivity
import android.os.Bundle
import androidx.core.graphics.drawable.toBitmap
import androidx.core.graphics.drawable.toDrawable
import com.wlp.myanjunote.R

/**
（1）Bitmap 是什么
Bitmap 是位图信息的存储，即⼀个矩形图像每个像素的颜⾊信息的存储器。

（2）Drawable 是什么
Drawable 是⼀个可以调⽤ Canvas 来进⾏绘制的上层⼯具。调⽤
Drawable.draw(Canvas) 可以把 Drawable 设置的绘制内容绘制到 Canvas
中。
Drawable 内部存储的是绘制规则，这个规则可以是⼀个具体的 Bitmap，也可以是
⼀个纯粹的颜⾊，甚⾄可以是⼀个抽象的、灵活的描述。Drawable 可以不含有具体
的像素信息，只要它含有的信息⾜以在 draw(Canvas) ⽅法被调⽤时进⾏绘制就
够了。
由于 Drawable 存储的只是绘制规则，因此在它的 draw() ⽅法被调⽤前，需要先
调⽤ Drawable.setBounds() 来为它设置绘制边界。

（3）Bitmap 和 Drawable 的互相转换
事实上，由于 Bitmap 和 Drawable 是两个不同的概念，因此确切地说它们并不是互
相「转换」，⽽是从其中⼀个获得另⼀个的对象：
Bitmap -> Drawable：创建⼀个 BitmapDrawable。
Drawable -> Bitmap：如果是 BitmapDrawable，使⽤BitmapDrawable.getBitmap() 直接获取；
如果不是，创建⼀个 Bitmap和⼀个 Canvas，使⽤ Drawable 通过 Canvas 把内容绘制到 Bitmap 中。

 （4）⾃定义 Drawable
1）怎么做？
重写⼏个抽象⽅法
重写 setAlpha() 的时候记得重写 getAlpha()
重写 draw(Canvas) ⽅法，然后在⾥⾯做具体的绘制⼯作
例如：MeshDrawable
2）有⽤吗？
有⽤。它就是⼀个更加抽象和专注的、仅仅⽤于绘制的⾃定义 View 模块。
3）⽤来⼲嘛？
需要共享在多个 View 之间的绘制代码，写在 Drawable ⾥，然后在多个⾃
定义 View ⾥只要引⽤相同的 Drawable 就好，⽽不⽤互相粘贴代码。
例如？
股票软件的多个蜡烛图界⾯，可以把共享的蜡烛图界⾯放进去
 **/
class MainActivity : AppCompatActivity() {
  override fun onCreate(savedInstanceState: Bundle?) {
    super.onCreate(savedInstanceState)
    setContentView(R.layout.activity_main)

    //互转：Bitmap --> drawable
    /*val bitmap = Bitmap.createBitmap(20, 20, Bitmap.Config.ARGB_8888)
    bitmap.toDrawable(resources)//bitmap --> BitmapDrawable 看看方法的源代码

    //drawable --> bitmap
    val drawable = ColorDrawable()
    drawable.toBitmap()//看看方法的源代码

    val bit = BitmapDrawable()
    bit.bitmap //getDrawable*/
  }
}